File: hw4/scala/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import java.util.Date
import scala.util.Random
import scala.util.control.Breaks

object task1 {

  def hash_func(x:Int, a:Int, b:Int, m:Int) = {
    (a * x.toInt + b) % m
  }

  def find_prime(x:Int): Int ={
    var isPrime = false
    var x_pri = x
    while (!isPrime) {
      x_pri += 1
      val loop = new Breaks;
      var tmp = 2
      loop.breakable {
        for( i <- 2 until x){
          if (x_pri % i == 0){
            loop.break()
          }
          tmp += 1
        }
      }
      if (tmp == x) {
        isPrime = true
      }
    }
    x_pri
  }

  def generate_hash_parameters(n_features:Int, total_rows: Int)= {
    val m = find_prime(n_features)
    val r = Random
    var a_list:List[Int] = List()
    var b_list:List[Int] = List()
    while (a_list.length != total_rows) {
      val a = r.nextInt(500) + 1
      val b = r.nextInt(5000) + 100
      if (a % m != 0 && !a_list.contains(a)) {
        a_list = a_list :+ a
        b_list = b_list :+ b
      }
    }
    (a_list, b_list, m)
  }

  def signature(item:List[Int], a_list:List[Int], b_list:List[Int], m:Int) = {
    var signature: List[Int] = List()
    for (i <- a_list.indices) {
      var min_signature: List[Int] = List()
      for (j <- item.indices) {
        val hash_val = hash_func(item.apply(j), a_list.apply(i), b_list.apply(i), m)
        min_signature = min_signature :+ hash_val
      }
      signature = signature :+ min_signature.min
    }
    signature
  }

  def split_band(pair:(String, List[Int]), r:Int, b:Int) = {
    var tmp:List[(String, String)] = List()
    val business = pair._1
    val signatures = pair._2
    for (i <- 0 until b) {
      val band = signatures.slice(r*i, r*(i+1))
      var key = i.toString + "_"
      band.foreach(item=>key += item.toString)
      tmp = tmp :+ (key, business)
    }
    tmp
  }

  def jaccard_similarity(x:List[String], business_users:Map[String, List[Int]]) = {
    val s1 = business_users.get(x.head)
    val s2 = business_users.get(x.apply(1))
    val similarity = s1.head.toSet.intersect(s2.head.toSet).size * 1.0 /  s1.head.toSet.union(s2.head.toSet).size
    (x, similarity)
  }

  def main(args: Array[String]): Unit = {
    val start_time =new Date().getTime
    val conf = new SparkConf().setMaster("local[*]").setAppName("task1")
    val sc = new SparkContext(conf)

    val input_file = args(0)//"E:\\Data\\inf553\\hw3\\publicdata\\train_review.json"
    val output_file = args(1)//"task1.scala.res"
    val total_rows = 40
    val band_rows = 1

    val rdd_review = sc.textFile(input_file).map(item => parse(item))
      .map(item => (compact(item \ "business_id"), compact(item \ "user_id")))


    val rdd_users = rdd_review.map(item=>item._2).distinct().sortBy(item=>item)
    val user_index_dict = rdd_users.zipWithIndex().collect().toMap

    val hash_parameters = generate_hash_parameters(user_index_dict.size, total_rows)
    val a_list = hash_parameters._1
    val b_list = hash_parameters._2
    val m = hash_parameters._3

    val rdd = rdd_review.groupByKey().mapValues(item=>{
      var tmp:List[Int] = List()
      item.foreach(i=>tmp = tmp :+ user_index_dict.get(i).toList.head.toInt)
      tmp
    })
    val rdd1 = rdd.mapValues(item=>signature(item, a_list, b_list, m))
    val rdd2 = rdd1.flatMap(item=>split_band(item, band_rows, total_rows/band_rows.toInt))

    val rdd3 = rdd2.groupByKey().map(item=>item._2.toList).filter(item=>item.length > 1).flatMap(item=>item.combinations(2)).map(item=>item.sorted).distinct()

    val business_users = rdd.collect().toMap

    val result = rdd3.map(item=>jaccard_similarity(item, business_users)).filter(item=>item._2 >= 0.05)

    val pw = new PrintWriter(new File(output_file))
    result.collect().foreach(item=>pw.write("{\"b1\": "+item._1.head + ", \"b2\": " +item._1.apply(1) + ", \"sim\": " + item._2 + "}\n"))
    pw.close()
    val end_time =new Date().getTime
    println("time: "+(end_time-start_time)/1000.0+"s")
  }


}

File: hw3/python/task1.py
import os
import sys
import json
import time
import random
from itertools import combinations

from pyspark import SparkContext


def hash_func(x, a, b, m):
    return (a * x + b) % m

def find_prime(x):
    isPrime = False
    while not isPrime:
        x = x + 1
        for i in range(2, x):
            if x % i == 0:
                break
        else:
            isPrime = True
    return x

def generate_hash_parameters(n_features, num_rows):
    m = find_prime(n_features)
    random_a = list()
    random_b = list()
    while len(random_a) != num_rows:
        a = random.randint(1, 500)
        b = random.randint(10, 50000)
        if a % m != 0 and a not in random_a:
            random_a.append(a)
            random_b.append(b)
    return random_a, random_b, m


def signature(users, a_list, b_list, m):
    signature = list()
    for i in range(len(a_list)):
        min_signature = min(list(map(lambda x: hash_func(x, a_list[i], b_list[i], m), users)))
        signature.append(min_signature)
    return signature


def split_band(pair, r, b):
    business, signatures = pair[0], pair[1]
    for i in range(b):
        band = signatures[r*i: r*(i+1)]
        yield "{}_{}".format(i, "_".join([str(i) for i in band])), business

def jaccard_similarity(x, business_users):
    s1, s2 = business_users[x[0]], business_users[x[1]]

    similarity = len(set(s1).intersection(s2))/ len(set(s1).union(s2))
    return x, similarity

if __name__ == '__main__':
    start_time = time.time()
    os.environ['JAVA_HOME'] = '/home/buyi/app/jdk'
    os.environ['PYSPARK_PYTHON'] = '/usr/bin/python3'
    #input_file, output_file = '/home/buyi/data/hw3/train_review.json', 'task1.res'
    input_file, output_file = sys.argv[1:]
    num_rows = 40
    # num_bands = 20
    r = 1
    sc = SparkContext('local', 'task1')

    rdd_review = sc.textFile(input_file).map(json.loads).map(lambda x:(x['business_id'],x['user_id'])).persist()
    rdd_users = rdd_review.map(lambda x: x[1]).distinct().sortBy(lambda x: x)#.map(lambda x:(x[1],x[0]))
    user_index_dict = dict(rdd_users.zipWithIndex().collect())
    # print("the number of users: {}".format(len(user_index_dict)))

    a_list, b_list, m = generate_hash_parameters(len(user_index_dict), num_rows)

    # print("m: {}".format(m))
    rdd = rdd_review.groupByKey().mapValues(lambda x:[user_index_dict[i] for i in x])
    # print("characteristic matrix: {}".format(rdd.take(5)))
    # print("time: {}".format(time.time() - start_time))
    rdd1 = rdd.mapValues(lambda x:signature(x, a_list, b_list, m))
    # print("signatures matrix: {}".format(rdd1.take(5)))

    rdd2 = rdd1.flatMap(lambda x:split_band(x, r, int(num_rows/r)))

    rdd3 = rdd2.groupByKey().map(lambda x:list(x[1])).filter(lambda x:len(x) > 1).flatMap(lambda x:list(combinations(x,2))).map(lambda x:tuple(sorted(x))).distinct()
    # print("candidate pairs: {}".format(rdd3.count()))
    business_users = dict(rdd.collect())
    result = rdd3.map(lambda x:jaccard_similarity(x, business_users)).filter(lambda x:x[1] >= 0.05)
    data = result.collect()
    # print(result.take(10))
    with open(output_file, 'w') as f:
        for item in data:
            tmp = {'b1':item[0][0], 'b2': item[0][1], 'sim': item[1]}
            f.write(json.dumps(tmp)+"\n")
    print("time: {}".format(time.time() - start_time))

